Disallow hyphens in Rust crate names

lfairy opened the pull request:
Disallow hyphens in Rust crate names, but continue allowing them in Cargo packages.

[Rendered](https://github.com/lfairy/rfcs/blob/hyphens-considered-harmful/text/0000-hyphens-considered-harmful.md)

carllerche commented:
:-1: Dashes are more readable and renames aren't the end of the world.

For example. I may have a crate: `syncbox-io` and using it would be:

```rust
extern crate "syncbox-io" as sio
```

Once there are dashes or underscores in the name, it's probably easier to alias to something shorter anyway for use in code.

sfackler commented:
Big :+1: here. People should try to make libraries they author as painless to use as possible, and forcing all consumers of a crate to rename it every time it's imported is not a great way to do that.

lfairy commented:
@carllerche

> Dashes are more readable

The readability/aesthetics argument is subjective, and not something everyone agrees on.

Really, the fact that all arguments for them come down to *they look nice* should sound at least a bit suspicious. Especially when there are objective arguments on the other side.

> renames aren't the end of the world

They aren't, of course. But as @sfackler says, making this kind of renaming mandatory is not the way to go. It'll be yet another special case for newcomers to learn, and introductory tutorials to explain.

Quite a few features of Rust -- C-style syntax, lifetime elision, `IntoIterator` -- are geared toward reducing the number of concepts a beginner has to learn up front. Disallowing hyphens follows in this trend: beginners won't have to learn renaming syntax just to generate JSON.

Renaming crates will be optional sugar, as it should be.

> Once there are dashes or underscores in the name, it's probably easier to alias to something shorter anyway for use in code.

Note that if we disallow hyphens, we'll be able to do this instead:

```rust
extern crate syncbox_io as sio;
```

which compares favorably to

```rust
use std::fmt::Write as WriteFormat;
```

In fact, hyphens are the *only* character we allow in crates which aren't also valid in identifiers. Without them, we wouldn't need quotes at all. I didn't include this change in the RFC, but that syntax would be possible in the future.

(And before someone brings up Go: in that language the string is a [path to the library][1], not a crate name. Note that Go doesn't allow hyphens in crate/package names either.)

[1]: https://golang.org/ref/spec#Packages

liigo commented:
+1

CloudiDust commented:
I'll admit that `foo-bar` looks better to me, but I still prefer removing `-`s from crate names for less exceptional cases in the language and better consistency across the ecosystem, so +1.

tshepang commented:
let's do this... :+1:

jmesmon commented:
I suppose another alternate is to allow `-` in idents

sfackler commented:
It's ambiguous - is `a-b` the ident `a-b` or the ident `a` minus the ident `b`?

jmesmon commented:
@sfackler sure, we'd also need to change parsing rules (that would be a single ident), but it is an alternative.

tshepang commented:
To avoid ambiguity, whitespace would be needed in order to have `-`  as an operator. Much code would break.

tshepang commented:
Is there any language that allows `-` as part of identifier?

lfairy commented:
@tshepang Lisp-family languages do, but they usually require spaces between tokens for this reason. I don't know of a C or ML-style language that does.

tshepang commented:
@lfairy I forgot about Lisp, thanks.

divius commented:
+1, using rustc-serialize is a bit weird at first

P1start commented:
:+1: Not only is being forced to rename certain crates very frustrating, but it’s also annoying to have to choose between `-` and `_` when choosing a crate name, and to remember which of `-` and `_` a particular crate uses. Removing one of them seems to be a good course of action to me, and `-` is the best candidate for removal in my opinion by far (removing `_` would be inconsistent with identifiers, having `-` requires renaming, and the other points already discussed).

lifthrasiir commented:
I agree to @P1start's rationale. Encoding currently has both `encoding_index_tests` and `encoding-index-singlebyte` (and so on), primarily because they are registered by different authors. Having one correct way would be appreciatd.

steveklabnik commented:
Big :+1: here.

alexcrichton commented:
We [talked about this RFC](https://github.com/rust-lang/meeting-minutes/blob/master/weekly-meetings/2015-03-10.md) in today's meeting. One of the points in favor of using hyphens, which I don't think has been stressed enough, is that there is a large amount of precedent for using `-` in "package names". For example github repositories, ruby gems (for namespacing), and npm all have a large and established convention of using a hyphen.

I am personally sympathetic with the hyphen convention, and I believe that going against the precedent-at-large to gain consistency is not worth it. I think that this RFC does a great job of laying out the alternatives (thanks!), and I would personally push for the fuzzy matching alternative to become the detailed design.

Of the two motivations, consistency and usability, fuzzy matching would solve the usability problem. There are probably still some more details to hash out precisely what happens, but overall I think it's the best direction to go in. We can continue the trend of hyphens set by other communities without sacrificing usability. In this situation we would probably continue to retain the quoting syntax for renaming if necessary, it just wouldn't be necessary if you used the canonical "extern crate" name for a crate.

steveklabnik commented:
I don't think fuzzy matching really solves the usability problem, it just
adds complexity.

alexcrichton commented:
@steveklabnik could you elaborate a little more on why you think it doesn't solve the usability problem? I'm curious what use cases you have in mind and what possible extensions could be added (not that they should be, just curious!).

The specific use-case I have in mind is what's spelled out in the RFC:

```rust
extern crate "rustc-serialize" as rustc_serialize;
// vs
extern crate rustc_serialize;
```

reem commented:
-1 for similar reasons as @carllerche.

Hyphenated names are often *meant* to be renamed. For instance, I usually do `extern crate "rustc_serialize" as serialize`, especially now that we have `$crate` for macros.

Another case is with distributing iron-related stuff, perhaps I want to distribute a set of crates under `iron-X` (right now everything is just X but that could change and there are other frameworks that will exist) that are meant to be imported `extern crate "iron-x" as x`.

steveklabnik commented:
@alexcrichton 

Without hypens, crate names are just identifiers. With hypens and fuzzy matching, now I have to explain why it's not just an identifier, and what those fuzzy matchers are, when instead, we could all just use identifiers and sidestep this whole mess.

It's not like there are a zillion packages right now, but I certainly would use a non-hyphenated package over a hyphenated one, if they provided simliar functionality. It's ugly hoop-jumping for no actual benefit, imho.

seanmonstar commented:
:+1: while you certainly can rename to your liking, you shouldn't be forced to.

CloudiDust commented:
@reem, I am thinking, would it be a good alternative to state in the guide lines that `-` is a "namespacing"/"tagging" symbol, and `_` is for separating words in a name? So if a project `foo` has some component named `bar baz`, then the correct naming would be `foo-bar_baz` not `foo-bar-baz` nor `foo_bar_baz`.

But then on second thought, I'd rather we permit `/` as the namespacing symbol, *and* forbid `-`.

@Everyone, would the above be a good alternative?

alexcrichton commented:
> Without hypens, crate names are just identifiers. With hypens and fuzzy matching, now I have to explain why it's not just an identifier, and what those fuzzy matchers are, when instead, we could all just use identifiers and sidestep this whole mess.
> 
> I guess that's the core of it: we can pile on rules to make hypens look just like underscores, or we can just use underscores and call it a day.

I definitely agree that adding any form of fuzzy matching would be adding more complexity. This RFC classifies today's behavior as both complex and hindering usability. I'd like to tease apart these concerns in this discussion. You previously said "I don't think fuzzy matching really solves the usability problem", but it sounds like you're mostly worried about the complexity a fuzzy match would be adding?

To be clear, I'm not disagreeing that a fuzzy match adds complexity, only that a fuzzy match does *not* solve the usability problem. I think @sfackler put it well in that today a `-` in a crate name **forces** users to rename-on-import, but a fuzzy match does not have this restriction: the "canonical name" would be hyphens translated to underscores.

> It's not like there are a zillion packages right now, but I certainly would use a non-hyphenated package over a hyphenated one, if they provided simliar functionality. It's ugly hoop-jumping for no actual benefit, imho.

I agree, and my motivation for pushing back on disallowing hyphens is that I like hyphens, not so much that I don't want to go migrate all the crates on crates.io :)

alexcrichton commented:
@CloudiDust 

> Everyone, would the above be a good alternative?

I believe that scheme is similar to what Ruby employs, although I'm less sure about Python or NPM. It does clash, however, with the `*-sys` convention as well as the number 1 crate on crates.io, `rustc-serialize` (just as some data points).

Valloric commented:
Please don't add name fuzzy matching! Better leave it as is than add a possible footgun to the language. The fuzzy matching idea is much too clever for what crate importing is supposed to be doing.

Forcing everyone to just use the underscore is a better idea IMO, but damn-nigh everything is better than fuzzy matching the name. It's way too magic.

CloudiDust commented:
@alexcrichton, we may fuzzy match `/` and `-` (note: both mandates renaming)  instead of fuzzy matching `-` and `_`. And we may guide the community towards using `/` instead of `-`. (Or if some package uses `-` to separate words in a name, it will change to use `_`. `bloom-filter` looks good, `bloom/filter` may not be what the author intends.)

Also, we may mimic how `lein` (Clojure's package manager) does things, by treating a namespace-less name `foo` as the same as `foo/foo`. This worked well for clojure, I think.

The mandatory renaming avoids name clashes nicely, without introducing the concept of package/crate namespacing into the compiler. (Besides `foo = foo/foo`, which can be a simple string rewrite.)

CloudiDust commented:
@Valloric, I can accept fuzzy matching *in the transition period*, but if we can complete the transition without fuzzy matching, it is better.

liigo commented:
nice!
On Mar 11, 2015 9:08 AM, "Richard Zhang" <notifications@github.com> wrote:

> @alexcrichton <https://github.com/alexcrichton>, we may fuzzy match / and
> - (note: both mandates renaming) instead of fuzzy matching - and _. And
> we may guide the community towards using / instead of -. (Or if some
> package uses - to separate words in a name, it will change to use _.
> bloom-filter looks good, bloom/filter may not be what the author intends.)
>
> Also, we may mimic how lein (Clojure's package manager) does things, by
> treating a namespace-less name foo as the same as foo/foo. This worked
> well for clojure, I think.
>
> The mandatory renaming avoids name clashes nicely, without introducing the
> concept of package/crate namespacing into the compiler. (Besides foo =
> foo/foo, which can be a simple string rewrite.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/940#issuecomment-78182058>.
>

CloudiDust commented:
Actually the RFC itself proposes using fuzzy matching in the transition period. (If we don't allow this, there would be no "transition", I realized.)

Also, there is no point in doing `foo = foo/foo`. `lein` does this because maven mandates separated group id and artifact id. Rust/Cargo/Crates.io don't have such concepts.

So I now prefer we replace `-` with `/`. Each package with `-` in its name would then have to decide which to use instead, `/` or `_`, depending on the meaning of `-`.

CloudiDust commented:
In the future, Cargo and Crates.io can treat the "namespace" parts of package names specially, like for grouping. And namespace-less packages like `foo` will be treated as "both a namespace and the nameless/main package inside that namespace" or as "a package in the global namespace". I prefer the former interpretation, as it describes nicely the relationship between `foo` and `foo/sys`.

And rustc doesn't have to know anything about package namespaces. It only needs to know that `/` can be in crate names but not in identifiers.

jmesmon commented:
@CloudiDust adding `/` as a namespacing symbol (I assume by this you mean changing the syntax of `use` to allow use of `foo/bar` without quotes) would be very much the same as allowing `-` as a "namespacing symbol" (both are binary operators)

And keeping `-` has the added benefit of allowing crate names to stay the same.

That said: I'm not keen on allowing something special that isn't an identifier after the crate name. I'm under the impression that this RFC would still exist even if we could do `use foo-bar as foo` (note: no quotes) today.

If people want to name things with `-` (or `/`), perhaps identifiers should just be changed to include them (causing spaces or other separating tokens to be required when doing division or subtraction)

CloudiDust commented:
@jmesmon, I am not proposing `foo/bar` to be legal, but `"foo/bar"` to be legal. The only change compared to today is `-` gets replaced by `/`.

Currently, `-` may be a namespace sigil or a word separator, so package names are inconsistently using `-` and `_`.

But it is clear that `/` is a namespace sigil and `_` is a word separator. So people will naturally use them consistently if `-` gets replaced by `/`.

If we change all `-` to `_`, it won't be clear which parts are acting as namespaces.

Cargo/Crates.io/rustc don't have to special case `/` (other than allowing it in package/crate names), but when they want to (in the future), they will know which sigil to look for.

chrish42 commented:
Yes please. +1.

eternaleye commented:
I'm in favor of banning `-` in crate names. I'll note a few things:

- `*-sys` are, in the vast majority, placeholders rather than fully-realized crates.
- Perl historically did the fuzzy matching thing, with `'` (single-quote) being able to stand in for `::`. Nobody likes this, except the Acme namespace. It was a mistake.
- Easy things should be easy, hard things should be possible. Importing code is easy; it shouldn't be made artificially hard by way of mandatory renaming - but renaming is a useful tool in the more complex cases.
- Allowing `_` to stand in for `-` silently means crates could be ambiguous on import, even without actual name collisions. This violates the principle of least surprise.
- I also dislike `/` as namespacing: Just like `-`, it's not valid in identifiers, and we'd get this exact same bikeshed to paint a second time. We already have `::` for namespacing, and it's not currently valid in crate-name context AIUI. Similar things should look similar.

CloudiDust commented:
@eternaleye, I choose `/` because it is a common namespacing character *and* it is not `::`, as package namespacing has nothing to do with code namespacing inside a crate. And rustc/cargo/crates.io don't *have to* special case `/` more than they special case `-` now.

People will need to do package namespacing anyway. And giving dual purposes to `_` can make names unclear in the future. (Which part is the namespace and which part is the "actual" name?)

For a package like `rustc-serialize`, it can be argued that this is actually a package "serialize" under the `rustc` namespace, and it is natual to rename it to `serialize` when importing (we want the "serialize" component of the "rustc" project). While having to rename `bloom-filter` to `bloom_filter` seems pointless.

Also I am thinking, if we later special case `/` (more) in rustc, we can have:

```rust
extern crate "foo/bar" as bar;
extern crate "foo/baz" as baz;

// Sugar of the above
extern crate {bar, baz} from foo;
```

Or something similar. And no change needs to happen to cargo/crates.io.

eternaleye commented:
Well, part of what I object to is "more than they special case `-` now" when this discussion is about whether it's already too special :P

Also, I've never been a fan of trailing `from` in languages - it means I need to do lookahead in order to have any idea what's actually in play.

Also, the truth of the matter is *we don't have namespacing*. If we do get namespacing (which I agree is a needful thing), it should actually be thought out as such rather than being retrofitted on top of a symbol originally added for aesthetics.

CloudiDust commented:
@eternaleye, I think I am not fine with special casing `-`, because `-` and `_` are quite similar in semantics. So choosing one or another is a very subjective matter, and leads to much inconsistency. OTOH `/` is clearly a namespacing character if it is paired with `_`. So at least it wins in the consistency department. :P

The `from` sugar is just one of many possiblities.

Adding namespaces to rust packages is a much discussed topic. And I am currently one in favor of "no mendatory multi-level namespacing" (or "every top level package creates a new namespace").

I think building namespacing support on top of `/` would be quite natural and backwards-compatiable. (The compiler doesn't have to know anything about namespacing currently, and support, if needed, can be added easily later, which is a big advantage.)

CloudiDust commented:
Also, we may special case `/` in this way:

For a crate named `C`, `extern crate "C"` would bound the crate to `C`'s last name component, that is:

```rust
extern crate "foo"; // the crate is bound to `foo`, the same as `extern crate foo;`
extern crate "foo/bar"; // the crate is bound to `bar`
extern crate "foo/bar/baz"; // the crate is bound to `baz`
```

This is simply not possible if we replace all `-`s with `_`s.

jimmycuadra commented:
One thing to think about when considering the precedent for hyphens in package names in other languages is that the difference between a hyphen and an underscore is confusing to new users, and because there is no enforcement of the scheme that hyphen means namespace and underscore means word break, not everyone follows it. As a result, you have lots of commonly used packages that go against this advice, and the result for developers is that you can't rely on the convention to remember how a package's name is spelled. An example of this in Ruby is the factory_girl gem, which has a Rails-specific extension called factory_girl_rails. If this were to have followed the conventions, it would be named factory_girl-rails, which is awkward looking and probably the reason the authors decided not to do it that way.

It's clear that there is a use case for the idea of a namespaced package, but by having no enforced scheme for it, people will just do whatever they want, which makes for a more confusing experience for people consuming packages.

I rather like @CloudiDust's suggestion of using a slash as the namespace separator and disallowing the hyphen entirely. When the only two non-alphanumeric characters are slash and underscore, I think it's much more clear which one is a namespace separator and which is a word boundary. I'm not sure it's a good idea to try to infer anything about crate structure from that namespacing though. I think it should probably just be a naming convention and nothing more.

tl;dr Other languages show there's a need for a distinction between namespace and word boundary in a package name, but we shouldn't necessarily follow their lead in how they implemented the ideas, since they cause inconsistency and confusion.

rkjnsn commented:
:+1: for banning `-` and requiring crate names to be valid identifiers 
:-1: for fuzzy matching

arthurprs commented:
:+1: it's invalid in code, so it shouldn't be allowed outside it. Most users don't know how to alias the crate and eventually resort to IRC to find out, this obviously sucks.

Fuzzy matching is asking for trouble but

mikedilger commented:
:+1: for Disallow hyphens in crate names
:-1: for Disallow hyphens in package names

CloudiDust commented:
I thought more about `/` last night and there were some disadvantages:

When declaring dependencies in Cargo.toml, `foo/bar` must be quoted:

```toml
[dependencies]
foo/bar = "0.0.1" # Error
"foo/bar" = "0.0.1" # Okay
```

which can be a bit strange when all other keys are not quoted.

Also, if we follow the new convention, almost all `foo_macros` crates should be renamed to `foo/macros`, which means they must be renamed when imported. For macro crates, this is added labor for almost nothing.

I am now not sure whether adding `/` would be worth it.

adfernandes commented:
:+1: for banning '-' and requiring crate names to be valid identifiers

:-1: for fuzzy matching

:-1: for having renamed be a "natural" way to use a crate

lfairy commented:
Oh wow, this blew up much more than I expected :open_mouth:

---

@CloudiDust There's also a bit of ambiguity on crates.io. For example, currently we can look up the versions of the package `maud` at <https://crates.io/crates/maud/versions>. But if we allow `/` in package names, this path will be ambiguous: does it list the versions of `maud`, or show the package named `maud/versions`?

`.` doesn't have the URL problem, but it still needs to be quoted in TOML:

```toml
[dependencies."maud.macros"]
git = "..."
```

So given these constraints, `-` and `_` are really the best options here.

I quite like your `extern crate` idea though. Perhaps we can use `-` as the separator instead of `/`? So we have this:

```rust
extern crate "rustc-serialize";
// desugars to:
extern crate "rustc-serialize" as serialize;
```

Go does something similar, which is a good sign. It does feel a bit icky, though:

* Nowhere else in the language do we introduce identifiers based on a string literal
* It doesn't make sense under the `*-sys` convention
* It's hard to see at a glance what the last component should be

So I'm hesitant about it.

---

@alexcrichton I wrote up a detailed argument against fuzzy matching, but now I'm not so sure...

For posterity, here are some points to consider:

* The majority of platforms allow hyphens, with the exceptions being C#, Java and Go. Rust tends not to go against the grain (pun not intended) without a very good reason.

* crates.io already rejects packages that differ only in capitalization (e.g. you can't upload a package called `RUSTC-SERIALIZE`). Identifying `-` and `_` would be a simple extension of this behavior.

* Mac and Windows are case-insensitive, so there's already a level of ambiguity involved.

* We can address the ambiguity issue by normalizing output filenames to lowercase-with-underscores. For example, the crate `Foo-BaR` can compile down to `libfoo_bar.so`. That seems like a good thing to do in general, hyphens or no.

Overall I'm starting to get tired of the debate over this thing, so I hope we can reach a consensus soon.

ncm commented:
1 File names are different from identifiers.  Always have been, always will be.

2 Organizations have internal standards for file naming that do not recognize Rust's existence.

3 It is perfectly reasonable to offer a fast track from file name to identifier when the file name conforms to identifier rules, e.g. no "", no 'as too' needed.

4 It is perfectly reasonable to require that non-conforming file names be quoted and aliased, i.e. "aa-bb.cc" as aabbcc;

5 It would be of service where names come from outside the Rust world to have relaxed restrictions, e.g. allow /-.+\[]()!?#@, and even space, but only inside quotes. It looks like a string literal, why impose unnecessary non-stringy restrictions?

6 Where the name is in a string literal, there's no point in folding anything; use the text as written.  The file system might fold names, and that is its business and none of Rust's.

7 In practice, stuff purely in the Rust world will have the names that can be expressed most conveniently, without coercion.  Folding case and separators might seem like a convenience, but is hardly worth the bother to document and for users to remember.

8 This boils down to 2 rules: 1) names not quoted are identifiers, full stop, and become file names by the identity mapping. 2) names in string literals, quoted, are (parts of) file names, not identifiers, full stop, and must be aliased to an identifier explicitly to work at all.

lfairy commented:
@ncm Crate names and filenames are orthogonal. You can set them differently using the `-o` and `--extern` options in rustc:

    # Build a library with a custom filename
    rustc --crate-type=lib --crate-name=mylibrary -o /path/to/libsuper-duper-custom-filename.rlib src/lib.rs

    # Link a library with a custom filename
    rustc --extern mylibrary=/path/to/libsuper-duper-custom-filename.rlib src/main.rs

aidancully commented:
I'll reiterate the [comment I made on internals](http://internals.rust-lang.org/t/pre-rfc-resolve-support-for-hyphens-in-crate-names/1459/22?u=aidancully): for Rust 1.0, the emphasis should be on preserving forward compatibility with desired future language evolution. Namespaces are, I think, too large a concept to design on the 1.0 schedule, but they *could* impact `extern crate` usage. Therefore, I think we should reduce `extern crate` to the simplest core that can possibly work for 1.0, so that we leave room to expand it later. To me, that means disallowing hyphens in crate names (since the way we have to rename crates now may not be the way we want to rename them in the future, hyphens are a forward-compatibility risk), turning off crate renaming entirely (for now), and possibly re-adding them after an explicit design process post-1.0.

In other words, I am actually OK with the current behavior post-1.0, so long as it's the result of an explicit design process. I think 1.0 itself, though, should use the simplest system that can work: no hyphens, no crate renames.

lfairy commented:
Okay -- I've rewritten the RFC to only disallow `-` in Rust crates, but continue to allow them in Cargo packages. Funnily enough, it ends up simpler than the original RFC as it doesn't need to migrate anything.

I also added a proposal to remove quotes from the `extern crate` syntax. I believe this fits with the goals of improving consistency and usability, and I don't see any drawbacks to doing it.

@alexcrichton Does this address your concerns?

alexcrichton commented:
@lfairy I'd be fine with this! I have a few clarifications though:

* This means that with a package name of `foo-bar` Cargo will pass `--crate-name foo_bar` to the compiler, right?
* In Cargo, you can name targets differently than the package itself, and I suspect we would want this to be an error, right?

    ```toml
    [lib]
    name = "foo-bar"
    ```
* As a data point, I just checked and we have 0 crates currently that only differ on `-` vs `_`, so this should be an easy restriction to add to crates.io.

steveklabnik commented:
Rubygems uses `-` as a means of indicating an extension, you use `_` in the names of gems.

steveklabnik commented:
Another thing to mention, https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md specifically says use `camel_case` for crate names.

lfairy commented:
@alexcrichton

> * This means that with a package name of `foo-bar` Cargo will pass `--crate-name foo_bar` to the compiler, right?
>
> * In Cargo, you can name targets differently than the package itself, and I suspect we would want this to be an error, right?

Yes and yes. I've reworded the section to clarify these points, thanks!

> * As a data point, I just checked and we have 0 crates currently that only differ on `-` vs `_`, so this should be an easy restriction to add to crates.io.

Great!

CloudiDust commented:
If we are doing this (allow `-` in package names but not crate names), then we need a convention for `-`. Basically, I think we should use it for namespacing/tagging as if we are using `/`, so, the following are considered correct:

`piston-some_component`, `sdl2-sys`, `regex-macros`, `bloom_filter`.

(Note: macro packages are to be renamed, but I can also accept leaving them as an exception as "foo macros" flows natually, in contrast "foo sys" doesn't.)

Also when we create a new package in cargo, I hope it can warn about names that may break convention, like `foo_sys`, and for names containing `-` like `foo-bar_baz`, cargo can warn: "hyphens are for namespacing and tagging, does this look natual to you: `foo/bar_baz`?"

lfairy commented:
@CloudiDust I agree that we need an official convention on this, but I think we should propose that in a separate PR.

ncm commented:
If there is a separate mapping between file and crate names is certainly doesn't need to be duplicated here.

Mapping "-" and "_" still seems like a wart. If it is always a suffix anyway, maybe "-" could map to "::" in identifier-land?

ncm commented:
Or, could a "-anything" suffix be simply omitted from the identifier? "-suffixes" don't seem to disambiguate anything.

CloudiDust commented:
@ncm, `::` is for intra-crate namespacing, so I don't think we should do the `-` -> `::`mapping. (And we should not be ignoring "-suffixes". Too magical.) However, you've given me a new idea.

@Ifairy @alexcrichton, what about retaining `-` in package names and map them to `/`s in crate names and make `extern crate foo/bar_baz` valid?

(Note, there are no quotes around `foo/bar_baz`, and the names would be bound to `bar_baz`, we are not doing funny things with a string literal, but `/` in a particular context.)

The advantage is that it is very clear hyphens/slashes are namespacing sigils. (And hyphens have to be mapped anyway.)

The disadvantage is `foo-sys` have to be renamed if there are multiple low-level ffi crates being imported. But this is no worse than the status quo.

arthurprs commented:
The current state of the RPC is reasonable, although I'd prefer to forbid "-" on both ends for the sake of consistency.

Does the core team expect to make a decision on this next week? I mean, the schedule is tight for the beta release.

CloudiDust commented:
@arthurprs, I agree forbidding `-` on both ends is more consistent.

@Everyone, I suppose one of the reasons to retain hyphens in package names is to avoid massive renaming on crates.io. And, hyphens *are* used for namespacing already.

I looked at crates.io and found that most packages I saw were (correctly) using `-` for namespacing and `_` for word separation.

While the convention has not been officially established, I'd be very surprised if `-` were not going to be the official namespacing sigil.

Mapping hyphens to underscores have one important advantage: it is simple. But it has one disadvantage: the namespacing information is lost in source code. If all crates become `foo_bar_baz`s in source code, then it kind of defeats the purpose of using `-` in package names.

So hyphens -> slashes feels natural to me.

Note: `extern crate foo/bar/baz` binding `foo/bar/baz` to `baz` is just like `use foo::bar::Baz` binding `foo::bar::Baz` to `Baz`.

jmesmon commented:
@CloudiDust which accepted RFC (or other documentation) specifies that `-` in crate names _must_ be used for namespacing instead of as a normal seperator?

Unless there is something, I view the discussion on namespacing crates as something that probably belongs in a separate RFC.

lfairy commented:
@CloudiDust Another disadvantage would be inconsistency: this `/` syntax isn't used anywhere else in the language, so it would feel strange in context. I'm hesitant to push such a change for this reason.

@arthurprs The RFC is written with a tight time frame in mind (all proposed changes are really easy to implement) so it shouldn't be an issue.

@jmesmon I think you're in agreement here. He already mentions that "the convention has not been officially established".

CloudiDust commented:
@Ifairy, I'm afraid that mapping hyphens to underscores would tempt people to go "I import with underscores anyway, so I'll always name the packages with underscores only." But this may be a minor problem and I am worrying too much.

Let's go with this RFC as written then.

liigo commented:
+1
On Mar 13, 2015 9:45 AM, "Aidan Cully" <notifications@github.com> wrote:

> I'll reiterate the comment I made on internals
> <http://internals.rust-lang.org/t/pre-rfc-resolve-support-for-hyphens-in-crate-names/1459/22?u=aidancully>:
> for Rust 1.0, the emphasis should be on preserving forward compatibility
> with desired future language evolution. Namespaces are, I think, too large
> a concept to design on the 1.0 schedule, but they *could* impact extern
> crate usage. Therefore, I think we should reduce extern crate to the
> simplest core that can possibly work for 1.0, so that we leave room to
> expand it later. To me, that means disallowing hyphens in crate names
> (since the way we have to rename crates now may not be the way we want to
> rename them in the future, hyphens are a forward-compatibility risk),
> turning off crate renaming entirely (for now), and possibly re-adding them
> after an explicit design process post-1.0.
>
> In other words, I am actually OK with the current behavior post-1.0, so
> long as it's the result of an explicit design process. I think 1.0 itself,
> though, should use the simplest system that can work: no hyphens, no crate
> renames.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/940#issuecomment-78740822>.
>

alexcrichton commented:
We [discussed this RFC](https://github.com/rust-lang/meeting-minutes/blob/master/weekly-meetings/2015-03-17.md#hyphens-and-crate-names) today and the reception was quite favorable of the RFC as-written. I'm going to hold off on merging for a bit to allow any final comments, but otherwise I think this is good to go, thanks @lfairy!

alexcrichton commented:
The current version of this RFC appears to have widespread support now by striking a nice balance between consistency in the language while allowing usage of naming conventions in external functions such as Cargo and crates.io. I've now merged this RFC, and thanks again @lfairy!

Tracking issue: https://github.com/rust-lang/rust/issues/23533

pnkfelix commented:
@alexcrichton was the intention here to also forbid hyphens in the "crate name" of an output binary?

It seems like we need to do something with our test suite to avoid lots of annoying warnings there.

alexcrichton commented:
@pnkfelix currently it is invalid to manually specify `--crate-name` with a hyphen in it, but compiling `foo-bar.rs` as a binary will cause the compiler to emit `foo-bar` as a binary with `foo_bar` as the crate name (similar to what Cargo does). This means that the test suite shouldn't have any extra warnings.

pnkfelix commented:
@alexcrichton all I know is that when I currently compile tests by hand via the command lines I see from `compiletest`, I often see warnings.

e.g.:
```
% x86_64-apple-darwin/stage2/bin/rustc /Users/fklock/Dev/Mozilla/rust-lesser-box/src/test/compile-fail/borrowck-issue-14498.rs -L x86_64-apple-darwin/test/compile-fail/ --target=x86_64-apple-darwin -L x86_64-apple-darwin/test/compile-fail/borrowck-issue-14498.stage2-x86_64-apple-darwinlibaux -C prefer-dynamic -o x86_64-apple-darwin/test/compile-fail/borrowck-issue-14498.stage2-x86_64-apple-darwin --cfg rtopt --cfg debug -O -L x86_64-apple-darwin/rt
warning: crate names soon cannot contain hyphens: borrowck-issue-14498
...
```

alexcrichton commented:
@pnkfelix are you sure you're rebased on the current master? I believe I removed that warning very recently.

pnkfelix commented:
@alexcrichton yeah okay I don't see it now.

gsingh93 commented:
Now that quotes are removed, are crate names like [static](https://github.com/iron/static) no longer valid?
